Add guideline about which dispatch keyword to use in native_functions.yaml.

ailzhang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #46354 Remove catchAllKernel_.
* #45939 Move catchAll to Math
* #46353 Update VariableTypeManual.cpp to not use catchAllKernel.
* #46352 Allow Undefined to get kernel from Math/DefaultBackend.
* **#46126 Add guideline about which dispatch keyword to use in native_functions.yaml.**
* #46236 [Reland] Update native_functions.yaml to add DefaultBackend.



Differential Revision: [D24233887](https://our.internmc.facebook.com/intern/diff/D24233887)

Commit:
Add guideline about which dispatch keyword to use in native_functions.yaml.

[ghstack-poisoned]

Commit:
Update on "Add guideline about which dispatch keyword to use in native_functions.yaml."


Differential Revision: [D24233887](https://our.internmc.facebook.com/intern/diff/D24233887)

[ghstack-poisoned]

Commit:
Update on "Add guideline about which dispatch keyword to use in native_functions.yaml."


Differential Revision: [D24233887](https://our.internmc.facebook.com/intern/diff/D24233887)

[ghstack-poisoned]

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+
+1. Think about inference, does your kernel work for all backends?
+
+    - No: you're likely providing different kerenls for different backends, e.g. DispatchStub or
```
kernels

ezyang replied:
BTW, the fact that DispatchStub doesn't work for all backends is non-obvious and probably worth explicitly calling out

bhosmer replied:
+1

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+1. Think about inference, does your kernel work for all backends?
+
+    - No: you're likely providing different kerenls for different backends, e.g. DispatchStub or
+      backend-depenent logics are used in the implementation.
```
dependent

ezyang replied:
also typically it's just "logic", no trailing s

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -279,13 +279,19 @@ to reuse the same function name in both cases.
-In addition to backends above, we also support keyword `Math` which is an alias
-that maps to all backend and autograd backend keys. In other words, function registered to `Math` key
-should be a plain mathematical composition of other `at::` functions and works for any backend.
+In addition to backends above, we also support keywords:
```
to the backends

ezyang replied:
support the keywords

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -279,13 +279,19 @@ to reuse the same function name in both cases.
+In addition to backends above, we also support keywords:
+  - `DefaultBackend`: an alias that maps to all backends. Functions registered to
+    `DefaultBackend` should work for any backend inference.
+  - `Math`: an alias that maps to all backend and autograd backend keys. Function
```
Functions

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -279,13 +279,19 @@ to reuse the same function name in both cases.
+    `DefaultBackend` should work for any backend inference.
+  - `Math`: an alias that maps to all backend and autograd backend keys. Function
+    registered to `Math` key should be a plain mathematical composition of other
+    `at::` functions and supports training and inference for any backend.
```
you are referring to the function here, which should be plural, which means supports here should just be "support"

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+### Choosing the right dispatch keyword
+
+After writing a native function in C++, it's important to think about which dispatch keyword
+to use in native_functions.yaml as it gives dispatcher information about backend and autograd support
```
the dispatcher

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+
+Here're steps to follow to decide the right dispatch keyword:
+
+1. Think about inference, does your kernel work for all backends?
```
colon instead of comma; comma here is a comma splice because the two sides are complete sentences

ezyang commented on aten/src/ATen/native/README.md:
````diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+        QuantizedCPU: kernel_quantized_cpu
+      ```
+
+      [END] Now this op will be called in `CPU/CUDA/QuantizedCPU` backend inference!
````
The meaning of `[END]` isn't really defined here; I'd probably suggest just using plain english, e.g., "Stop" or "You're done"

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+
+      [END] Now this op will be called in `CPU/CUDA/QuantizedCPU` backend inference!
+
+      Note if you want this op to support training as well, you're required to write a formula in
```
Note:

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+
+      [END] This will allow this op to be correctly registered for both inference and training.
+
+    - Yes, but you still want to provide a numerically stable gradient formula instead of using autorgrad, write
```
autograd

ezyang commented on aten/src/ATen/native/README.md:
```diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+      [END] Now this op is only used in inference. Note that now you're required to add a autograd formula,
+      or it'll error out in forward pass when calling with a Tensor has requires_grad=True.
+
+    - No: ops in this category are mainly `_out` functions that don't have a derivative formula defined.
```
It might be worth explicitly adding the term "boilerplate" here. It's not so easy to see what exactly you mean by `_out` here; in particular, you are talking about a function that is calling an `_out` function, not an `_out` function itself.

bhosmer replied:
+1

ezyang commented on aten/src/ATen/native/README.md:
````diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+      ```
+
+      [END] Now this op is only used in inference. Note that now you're required to add a autograd formula,
+      or it'll error out in forward pass when calling with a Tensor has requires_grad=True.
````
Not true; it's a delayed error right? Only error if you try to backwards on it.

ezyang commented:
This isn't high stakes README so the grammar fixes are not that important, but I hope the editing may be more generally useful for you.

bhosmer commented on aten/src/ATen/native/README.md:
```diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+  - direct registration to backend always has higher precendence than alias
+  - DO NOT provide multiple alias keywords to the same op: alias keywords have precedence `DefaultBackend > Math`,
+    e.g. adding both `Math` and `DefaultBackend` kernels for one op will completely ignore `Math` kernel for
+    both inference and training.
```
I thought we asserted on this case?

ailzhang replied:
Yup we have an assert in this case, adding it here just in case developers need to understand why a DefaultBackend is needed. 
Also just as a note here: currently we don't fail if users directly register to Math and DefaultBackend using a C++ API (this is unlikely but possible).

bhosmer commented on aten/src/ATen/native/README.md:
````diff
@@ -384,6 +390,72 @@ will be automatically differentiated! This can be the case if the function imple
+      ```
+
+      [END] Now this op is only used in inference. Note that now you're required to add a autograd formula,
+      or it'll error out in forward pass when calling with a Tensor has requires_grad=True.
````
Echoing Ed's note from line 446 here

bhosmer reviewed (approved):
LG! Couple suggestions, mostly just echoing Ed's (see inline +1s)

Commit:
Update on "Add guideline about which dispatch keyword to use in native_functions.yaml."


Differential Revision: [D24233887](https://our.internmc.facebook.com/intern/diff/D24233887)

[ghstack-poisoned]

Commit:
Update on "Add guideline about which dispatch keyword to use in native_functions.yaml."


Differential Revision: [D24233887](https://our.internmc.facebook.com/intern/diff/D24233887)

[ghstack-poisoned]

Commit:
Update on "Add guideline about which dispatch keyword to use in native_functions.yaml."




Differential Revision: [D24233887](https://our.internmc.facebook.com/intern/diff/D24233887)

[ghstack-poisoned]

Commit:
Update on "Add guideline about which dispatch keyword to use in native_functions.yaml."




Differential Revision: [D24233887](https://our.internmc.facebook.com/intern/diff/D24233887)

[ghstack-poisoned]

Commit:
Update on "Add guideline about which dispatch keyword to use in native_functions.yaml."




Differential Revision: [D24233887](https://our.internmc.facebook.com/intern/diff/D24233887)

[ghstack-poisoned]

Commit:
Update on "Add guideline about which dispatch keyword to use in native_functions.yaml."




Differential Revision: [D24233887](https://our.internmc.facebook.com/intern/diff/D24233887)

[ghstack-poisoned]
